agent-control/src/sub_agent/on_host/supervisor.rs

```diff
@@ -212,152 +208,141 @@ impl NotStartedSupervisorOnHost {
+                        if let Err(err) = health_publisher.publish((
+                            exec_id.to_string(),
+                            HealthWithStartTime::new(unhealthy.into(), supervisor_start_time),
+                        )) {
+                            error!("Publishing health status: {err}",);
+                        }
+
+                        0 // Default exit code
+                    }
+                };
+
+                if stop_consumer.is_cancelled() {
+                    info!(supervisor = bin, msg = "Executable terminated");
+                    break;
+                }
+
+                // check if restart policy needs to be applied
+                if !restart_policy.should_retry(exit_code) {
+                    info!(msg = "Executable not running");
+
+                    warn!(
+                        "Executable won't restart anymore due to having exceeded its restart policy"
                     );
+
                     debug!(
-                        "Informing of executable as unhealthy as there was an error launching it"
+                        "Informing of executable as unhealthy because the restart policy was exceeded"
+                    );
+                    let unhealthy = Unhealthy::new(
+                        "executable exceeded its defined restart policy".to_string(),
                     );
-                    let unhealthy = Unhealthy::new(format!("Error launching process: {err}"));
                     if let Err(err) = health_publisher.publish((
-                        executable_data_clone.id.to_string(),
+                        exec_id.clone(),
                         HealthWithStartTime::new(unhealthy.into(), supervisor_start_time),
                     )) {
-                        error!(
-                            "Error publishing health status for {}: {}",
-                            executable_data_clone.id, err
-                        );
+                        error!("Publishing health status: {err}");
                     }
-                })
-                .map(|exit_status| {
-                    handle_termination(
-                        &executable_data_clone,
-                        exit_status,
-                        health_publisher.clone(),
-                        &agent_id,
-                        &executable_data_clone.bin,
-                        supervisor_start_time,
-                    )
-                });
+                    break;
+                }
 
-            // A context cancelled means that the supervisor has been gracefully stopped and is the
-            // most probably reason why process has been exited.
-            if *Context::get_lock_cvar(&ctx).0.lock().unwrap() {
                 info!(
-                    supervisor = executable_data_clone.bin,
-                    msg = "supervisor has been stopped and process terminated"
+                    "Restarting supervisor ({}/{})",
+                    i + 1,
```

Comment:
I know this is a very stupid comment, but I'd prefer to have this counter logged after it has been actually incremented, to save us the typical troubleshooting of the numbers not matching until we notice it was not updated or updated two times or something like that xD

Reply:
That's enough for me!